### DISCUSSION TITLE
Sapling-like revset navigation

### DISCUSSION THREAD

**Discussion by mbStavola** [jj-vcs/jj#2038] [Show and tell]
One of the things I really enjoyed from Sapling was the ability to just say `sl prev` and `sl next` to move around. I was pretty bummed when I didn't see an equivalent in `jj`... but then I discovered the awesome power of revsets and the _amazing_ `revset-alias` config option. Now I'm happy as can be 😄 

Anyway, just thought I'd share for those who were looking for something similar:

```toml
[revset-aliases]
'prev' = '@-'
'next' = '@+'
'top' = 'latest(@:)'
'bottom' = 'latest(:(~:@) & :@)+ & :@'
```

Since they're just revsets you can use them in most contexts like so:

```
jj edit prev
jj co next
jj log -r bottom..top
```

**Comment by thoughtpolice** [comment]
Note that analogues for `prev` and `next` will be implemented in #1200; hopefully it'll go in sometime soon!

**Reply by mbStavola** [reply]
Ah! Totally missed that one, really cool to see this becoming a native feature of `jj`.

Still glad I messed around with this, if only to learn more about revsets 😄

**Comment by ilyagr** [comment]
The 'top` and 'bottom' aliases are neat!

I'm guessing you usually `jj edit` commits. I use slightly different aliases since I usually `jj new` commits and then `jj amend`. So, I replace `@` with `@-`. Unfortunately, I have to also do some extra gymnastics so that the current commit doesn't affect the outcome:

```toml
[revset-aliases]
'prev' = '@--'
'next' = '@-+~@'
'top' = 'latest(@-:~@:)'
'bottom' = ? # haven't translated this one yet, might update this later.
```

(I've used the first two for a while, I'll start using the `top` and `bottom` now that you showed them to me).

I'd love to build these aliases in; I wonder if there's some way to do it so that both workflows are supported. One option is to have two versions of each. However, that's not very elegant, and I can't come up with decent names for these.

**Reply by arxanas** [reply]
- git-branchless supports `git prev`/`git next` with the `-a`/`--all` flags to avoid having an extra `top`/`bottom` alias. But that only works if there's a notion of the "current commit stack" or "main branch".
- It would be nice to have a `stack()` revset so that `top()` is just `roots(stack())` and `bottom()` is `heads(stack())`.

**Reply by mbStavola** [reply]
> The 'top` and 'bottom' aliases are neat!

Thank you, they were also fun to figure out! I'm still thinking that `bottom` can be simplified somehow, but I don't quite see it just yet.

> I'm guessing you usually `jj edit` commits. I use slightly different aliases since I usually jj new commits and then jj amend.

_Usually_ is giving me a bit too much credit 😛-- I really just started playing with `jj` last night so I'm still figuring out what works for me and what my preferences are. You're correct in that I've been heavily using `jj edit`, but now I'm wondering if `new` + `amend` may actually make more sense for me. Just to update my mental model a bit, can you tell me why you prefer one over the other?

> I'd love to build these aliases in

That would be _awesome_.

> I wonder if there's some way to do it so that both workflows are supported.

Kinda dirty but, you could add a revset function like `relativize_to_command()` which would add the command's intended "target" to the passed in revset. That is, depending on the command it may append an operator to the revset passed in.

For example, consider what the commands _target_. The `edit` command will select the exact revset you specify whereas the `new` command will always create a new commit one after the specified revset. You can think of this being roughly `x` and `x-`[0]. 

So, given:

```toml
[revset-aliases]
'prev' = 'relativize_to_command(@-)'
```

`jj edit prev` would evaluate to `@-` and `jj new prev` would evaluate to `@--`... or something like that. Maybe that's a bit too magic for revset functions, but it's one possible solution.

---

[0]: though not exactly, because it's going to be a _new_ revision not an existing one but I digress
